docs/design/uptime-reputation-calculation.md

```diff
@@ -0,0 +1,122 @@
+
+This design doc proposes a different way for calculating uptime check scores.
+
+## Background
+
+Currently our disqualification system is described in the [disqualification design doc](disqualification.md) which is based in the calculation of different _SN_ reputation scores as described in [node-selection design doc](node-selection.md).
+
+The problem with the current uptime _SN_ reputation score calculation is that it isn't based on the actual time that a node is being offline, by cause of to the nature of the chosen model, causing that an _SN_ will get disqualified due its number of uptime checks success and failure which implies that new nodes can get disqualified having less uptime check failures than veteran nodes.
+
+While we understand that an _SN_ that has remained in the network with a low uptime check failures could be considered a more reliable _SN_ than new ones which recently joined, we see, at the same time, that new _SN_ shouldn't be harsher penalized than veteran nodes due to uptime check failures because:
+
+1. Any _SN_ (veteran and new ones) requires some time to be offline for maintenance and both should have the same maximum amount of time for it.
+1. The end goal of the uptime checks is to ensure that the network has certain availability and retrievability and uptime check failures affects to them no matter on how long the _SN_ has remained in the network and how much data an _SN_ has, hence node with more or less data must be treated equally.
+
+This design doc is focused on how to design and implement a fair, simple and understandable uptime reputation calculation after our initial approach didn't behave as expected, causing the disqualification of several nodes without having a proof that they have failed enough uptime checks that the _SNs_ were offline long time enough which put in danger the minimum level of the network availability and retrievability.
+
+## Design
+
+The new process for doing the uptime checks is based on once an _SN_ is selected for having the check, if the _SN_ fails the initial check, then it's tracked to periodically receive uptime checks until (a) it successfully passes an uptime check or (b) gets disqualified due to many uptime check failures in one month.
+
+Once an _SN_ has failed the first uptime check, the _SN_ is considered offline and the subtotal time, which the _SN_ is considered offline, is the time elapsed between the first and the last failed uptime check. The total offline time of an _SN_ is the accumulation of the subtotals that the _SN_ has had during the ongoing month and when such total surpasses the maximum permitted percentage time, the _SN_ will get disqualified.
```

Comment:
Is "ongoing month" calculated based on the proceeding 30 days (whatever they were) or is it from the beginning of the calendar month?

If it's the latter, a node can decide, near the end of the month, to get some free "offline time". In other words, if there's only 8 hours left in the calendar month, and you need 10 hours to be disqualified, you can just go offline.

Reply:
> Is "ongoing month" calculated based on the proceeding 30 days (whatever they were) or is it from the beginning of the calendar month?

I had the same doubt, but after some thoughts, I thought that a calendar month is clearer from the _SN_ point of view and when I found metrics on some services they speak in months, which I interpreted as calendar months.

> If it's the latter, a node can decide, near the end of the month, to get some free "offline time". In other words, if there's only 8 hours left in the calendar month, and you need 10 hours to be disqualified, you can just go offline.

~Sure, but doesn't apply the same for any SLA commitment?~

As far as I understand the problem here is that a bunch of SNs which have 8 hours left at the end of the month may decide to go offline all at once and they won't get disqualified but the network may be broken.

We may randomize the period of time for each one as a solution.

Reply:
I think trailing 30 days is less gameable, and I personally think users would equally understand metrics like "You had 1.6 hours of downtime in the last 30 days" and "You had 1.6 hours of downtime this month", but I guess that's a product question.

Reply:
I believe the intention was a 30 day window.

Reply:
As commented in the meeting I will update the document to calculated in days.
It will also be mentioned that his value is configurable.